### PR TITLE
Add TaskManager deployment through Deployment resource

### LINKED ISSUE #409: Taskmanager ephemeral Deployment instead of Statefulset
We performed some experiments to test FlinkCluster creation time for jobs with a lot of replicas.

Experiments were performed on GKE k8s version `1.21.11-gke.900`
TaskManager details: `600 replicas`
- `cpu: 12`
- `mem: 96Gi`
- `storage: 600Gi`

For my experiments, I used the following definitions:
- [statefulset](https://gist.github.com/live-wire/41a74fe0a4b01783208cd77bf421df31)
- [deployment](https://gist.github.com/live-wire/0720bf2b59164fec3d1d8bc9be7daa2f)

---

Metric: `Startup time = All pods ready`

Available options :
1. Already available Statefulsets with PVCs.
    - `Startup time = 30+ minutes`
    - Pros:
      - Uses PVCs so we can specify storage request per pod and one volume is only mounted on one pod at a time.
      - Since state per indexed taskmanager is saved, recovery from a lost pod(TM) is possible
    - Cons:
      -  Slow. The way statefulset pods are spawned, is inherently sequential and slow. (Even with [Parallel podManagementPolicy](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#parallel-pod-management))
      
2. Deployments with [generic ephemeral volumes](https://kubernetes.io/docs/concepts/storage/ephemeral-volumes/#generic-ephemeral-volumes)
    - `Startup time = <6 minutes`
    - Pros:
      - Uses PVCs internally, so we can specify storage request per pod and one volume is only mounted on one pod at a time.
      - Fast. Replicaset pods are spawned in parallel and is quite fast even for large number of replicas with attached volumes.
    - Cons:
      - Because of the ephemeral nature of the storage, if a pod is lost, so is any state associated with its volume.
3. Deployments with [emptyDir](https://kubernetes.io/docs/concepts/storage/volumes/#emptydir) 
    - `Startup time = 70 seconds`
    - Pros: very fast startup
    - Cons: 
       - Uses node boot disk and can't provide storage requests per pod
       - Pods share the same disk on the node and there is no separation 
       

It is worthwhile to have support for ephemeral deployment instead of statefulset for Taskmanagers in the operator (at-least for batch workloads).

### PR DESCRIPTION
**Pull request by live-wire**
This change will allow a significantly faster startup time for batch workloads with high number of replicas
- Should resolve #409

### CODE REVIEW

File: apis/flinkcluster/v1beta1/flinkcluster_types.go [hunk 1]
```diff
@@ -336,6 +345,11 @@ type TaskManagerSpec struct {
 	// [More info](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#persistentvolumeclaims)
 	VolumeClaimTemplates []corev1.PersistentVolumeClaim `json:"volumeClaimTemplates,omitempty"`
 
+	// _(Optional)_ Defines the mounted volume's storage type. If not specified, the default is `Persistent`.
+	// Persistent storage type is used with Statefulsets for stateful recovery.
+	// Ephemeral storage type is used with Deployments for faster startup.
+	StorageType StorageType `json:"storageType,omitempty"`
```
[the remark this answers is not in the document]
Comment (regadas) [reply apis/flinkcluster/v1beta1/flinkcluster_types.go:351]: The thing is that you can use that with StatefulSet's sooo ... would it be better actually to just use a deployment type property? I'm leaning towards this actually.

File: controllers/flinkcluster/flinkcluster_observer.go [hunk 2]
```diff
@@ -501,6 +499,48 @@ func (observer *ClusterStateObserver) observeJobManagerStatefulSet(
...
+			observedClusterState.tmState = TaskManagerState{
+				storageType:   storageType,
+				tmStatefulSet: observedTmStatefulSet,
+				tmDeployment:  nil,
+			}
+		}
+	}
+
+	var observedTmDeployment = new(appsv1.Deployment)
+	err := observer.observeTaskManagerDeployment(observedTmDeployment)
```
Comment (jto) [inline controllers/flinkcluster/flinkcluster_observer.go:525]: This should be in the else branch of the `if storageType == v1beta1.StorageTypePersistent` check ?

File: controllers/flinkcluster/flinkcluster_observer.go [hunk 3]
```diff
@@ -501,6 +499,48 @@ func (observer *ClusterStateObserver) observeJobManagerStatefulSet(
...
+			log.Info("Observed TaskManager StatefulSet", "state", "nil")
+			observedTmStatefulSet = nil
+		} else {
+			log.Info("Observed TaskManager StatefulSet", "state", *observedTmStatefulSet)
+			observedClusterState.tmState = TaskManagerState{
+				storageType:   storageType,
+				tmStatefulSet: observedTmStatefulSet,
+				tmDeployment:  nil,
+			}
+		}
```
Comment (jto) [inline controllers/flinkcluster/flinkcluster_observer.go:521]: :art: This should maybe be extracted in its own function. The code is getting hard to follow with all those branches.

File: controllers/flinkcluster/flinkcluster_observer.go [hunk 4]
```diff
@@ -501,6 +499,48 @@ func (observer *ClusterStateObserver) observeJobManagerStatefulSet(
...
+			}
+			log.Info("Observed TaskManager StatefulSet", "state", "nil")
+			observedTmStatefulSet = nil
+		} else {
+			log.Info("Observed TaskManager StatefulSet", "state", *observedTmStatefulSet)
+			observedClusterState.tmState = TaskManagerState{
+				storageType:   storageType,
+				tmStatefulSet: observedTmStatefulSet,
+				tmDeployment:  nil,
+			}
```
Comment (jto) [inline controllers/flinkcluster/flinkcluster_observer.go:520]: :art:  This should maybe be extracted in its own function. The code is getting hard to follow with all those branches.

File: controllers/flinkcluster/flinkcluster_converter.go [hunk 5]
```diff
@@ -534,6 +546,60 @@ func newTaskManagerStatefulSet(flinkCluster *v1beta1.FlinkCluster) *appsv1.State
...
+
+// Gets the desired TaskManager Deployment spec from a cluster spec.
+func newTaskManagerDeployment(flinkCluster *v1beta1.FlinkCluster) *appsv1.Deployment {
+	var taskManagerSpec = flinkCluster.Spec.TaskManager
+	var taskManagerDeploymentName = getTaskManagerDeploymentName(flinkCluster.Name)
+	var podLabels = getComponentLabels(flinkCluster, "taskmanager")
+	podLabels = mergeLabels(podLabels, taskManagerSpec.PodLabels)
+	var deploymentLabels = mergeLabels(podLabels, getRevisionHashLabels(&flinkCluster.Status.Revision))
+
+	mainContainer := newTaskMangerContainer(flinkCluster)
```
Comment (jto) [inline controllers/flinkcluster/flinkcluster_converter.go:578]: There's a typo in that name

File: apis/flinkcluster/v1beta1/flinkcluster_types.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (live-wire) [reply apis/flinkcluster/v1beta1/flinkcluster_types.go:351]: Yes. Will be more clear to just use Deployment/Statefulset in the newly introduced property. 

In case of `kind: Deployment` however, I still want to take the volumeClaims from the TaskManagerSpec and append them to Ephemeral volumes in the Deployment. This way it will be easier for users to switch between the two `kinds`.

Wdyt?

File: controllers/flinkcluster/flinkcluster_controller.go [hunk 6]
```diff
@@ -211,11 +211,18 @@ func (handler *FlinkClusterHandler) reconcile(ctx context.Context,
...
 	}
-	if desired.TmStatefulSet != nil {
-		log.Info("Desired state", "TaskManager StatefulSet", *desired.TmStatefulSet)
+	if desired.TmDesiredState != nil {
+		if desired.TmDesiredState.StorageType == v1beta1.StorageTypePersistent && desired.TmDesiredState.StatefulSet != nil {
+			log.Info("Desired state", "TaskManager StatefulSet", *desired.TmDesiredState.StatefulSet)
+		} else if desired.TmDesiredState.StorageType == v1beta1.StorageTypeEphemeral && desired.TmDesiredState.Deployment != nil {
+			log.Info("Desired state", "TaskManager Deployment", *desired.TmDesiredState.Deployment)
+		} else {
+			log.Info("Desired state", "TaskManager", "nil")
```
[the remark this answers is not in the document]
Comment (live-wire) [reply controllers/flinkcluster/flinkcluster_controller.go:220]: It will be deleted in that case. Which seems like an *ok* thing to do (looking at other objects around it)

File: internal/model/model.go [hunk 7]
```diff
@@ -21,12 +21,18 @@ import (
...
+	StatefulSet    *appsv1.StatefulSet
+}
+
 // DesiredClusterState holds desired state of a cluster.
 type DesiredClusterState struct {
 	JmStatefulSet       *appsv1.StatefulSet
 	JmService           *corev1.Service
 	JmIngress           *networkingv1.Ingress
-	TmStatefulSet       *appsv1.StatefulSet
+	TmDesiredState      *TaskManagerDesiredState
```
Comment (regadas) [inline internal/model/model.go:35]: Since there are some changes planned around the observer/reconciler here I would keep things simple and just a `TmDeployment` instead of introducing this new type; No need to add an additional `DeploymentType` to the state since `TmDeployment` and `TmStatefulSet` will be mutually exclusive.

With this, we keep basically all the logic around StatefulSet (observer/reconciler) and just add the necessary bits for the Deployment.

If I'm seeing things right:
```
getDesiredClusterState(observed *ObservedClusterState) *model.DesiredClusterState
```

This should be the only place where we check `TaskManager.DeploymentType` and set either one.


This should simplify things a little bit and will make things easier for what comes next.

Comment (regadas) [inline internal/model/model.go:35]: Since there are some changes planned around the observer/reconciler here I would keep things simple and just a `TmDeployment` instead of introducing this new type; No need to add an additional `DeploymentType` to the state since `TmDeployment` and `TmStatefulSet` will be mutually exclusive.

With this, we keep basically all the logic around StatefulSet (observer/reconciler) and just add the necessary bits for the Deployment.

If I'm seeing things right:
```
getDesiredClusterState(observed *ObservedClusterState) *model.DesiredClusterState
```

This should be the only place where we check `TaskManager.DeploymentType` and set either one.


This should simplify things a little bit and will make things easier for what comes next.

File: apis/flinkcluster/v1beta1/flinkcluster_types.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (regadas) [reply apis/flinkcluster/v1beta1/flinkcluster_types.go]: Yeah I guess we can do that; basically appending to the volumes; We need to be explicit about that in the docs though.

[the remark this answers is not in the document]
Comment (regadas) [reply apis/flinkcluster/v1beta1/flinkcluster_types.go]: Yeah I guess we can do that; basically appending to the volumes; We need to be explicit about that in the docs though.

File: internal/model/model.go (hunk 7 above)
Comment (live-wire) [reply internal/model/model.go:35]: That would be cleaner indeed. Will do 👍

File: apis/flinkcluster/v1beta1/flinkcluster_types.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (live-wire) [reply apis/flinkcluster/v1beta1/flinkcluster_types.go]: Will add more comments next to volumeClaim too 👍

**Comment by jto** [comment]
This looks great! Could you had a couple of tests for the new type of deployment ?

**Comment by live-wire** [comment]
I'm done with the suggested changes.
Any more improvements @jto @regadas ?

**Review by regadas — CHANGES_REQUESTED** [review]
Looking good! just a few comments and I think we are good to go.

File: controllers/flinkcluster/flinkcluster_util.go [hunk 8]
```diff
@@ -122,6 +122,11 @@ func getTaskManagerStatefulSetName(clusterName string) string {
 	return clusterName + "-taskmanager"
 }
 
+// Gets TaskManager StatefulSet name
```
[the remark this answers is not in the document]
Comment (live-wire) [reply controllers/flinkcluster/flinkcluster_util.go]: oops

**Comment by live-wire** [comment]
Done! @regadas 
Let's 🚢 it?

**Review by regadas — APPROVED** [review]
